Bolts/Common/BFTask.m

```diff
@@ -418,16 +455,16 @@ - (void)waitUntilFinished {
     if ([NSThread isMainThread]) {
         [self warnOperationOnMainThread];
     }
-
-    @synchronized(self.lock) {
-        if (self.completed) {
-            return;
+    BOOL completed = self.isCompleted;
+    dispatch_barrier_sync(_synchronizationQueue, ^{
```

Comment:
We're locking twice here by going through the property (use ivar instead?). Theres also a TOCTOU race condition here.